text/0001-script-based-tokenizer.md

```diff
@@ -0,0 +1,187 @@
+- Algolia:
+	- Algolia seems to use default word segmentation use user configurable separators. The use can also set the `queryLanguage` or provide a list of `stopWords`, but it is unclear how this work under the hood for languages such as Mandarin. They also word decompounging, meaning that they can split words that are mode of multiple words (mostly germanic/nordic languages apparently).
+
+### Explanation
+
+The tokenizer serves as a proxy for other tokenizers, specialized in the language detected by the Tokenizer. It is instantiated with a String, and is then polled for tokens until it's delepted:
+```rust
+let tokenizer = Tokenizer::new("The quick brown fox jumps over the lazy dog");
+assert_eq!(tokenizer.next().txt(), "The")
+assert_eq!(tokenizer.next().txt(), " ")
+assert_eq!(tokenizer.next().txt(), "quick")
+```
+
+### Impact on documentation
+
+This feature should not impact the meilisearch user documentation,
+in future versions we will probably provide a way to configure tokenizer and this will be discussed in a new specification.
+
+## Technical Specifications
+
+### Architecture
+
+As the actual tokenizer, this tokenizer will be a standolone library,
+and will be used by meilisearch like another one.
+We decided to put it in an other repository to be able to use it in meilisearch/Meilisearch and in the future core-engine.
+
+### Implementation Details
+
+We could have 2 approach of tokenization based on the script,
+the main difficulty is to handle multi-script documents
+which need to segment each part of script and tokenize them differently,
+theses 2 approach are:
+- first iterate over the whole document, splitting it by script, and after, tokenize each part with specialized lexer
+- detect the script of the start of the document, tokenize with specialized lexer while script is the same, switch lexer if the script change
+
+#### Tokenizer
+
+```rust
+use crate::token::Token;
+use crate::internal_tokenizer::InternalTokenizer;
+
+struct Tokenizer<'a> {
+    /// script specialized tokenizer, this can be switched during
+    /// document tokenization if the document contains several scripts
+    current_tokenizer: Option<Box<dyn InternalTokenizer<'a>>>,
+    /// current character index in the document
+    current_char_index: u64,
+    /// reference on the document content
+    inner: &'a str,
+}
+
+impl<'a> Tokenizer<'a> {
+    /// create a new tokenizer detecting script
+    /// and chose the specialized internal tokenizer
+    fn new(inner: &'a str) -> Self { unimplemented!() }
+}
+
+impl<'a> Iterator for Tokenizer<'a> {
+    type Item = Token<'a>;
+
+    /// return the next token calling the `next` method of the internal tokenizer,
+    /// if the internal tokeizer return None, the function try to redetect script and chose a new tokenizer,
+    /// if no iternal tokenizer is chosen, the method return None
+    fn next(&mut self) -> Option<Self::Item> { unimplemented!() }
```

Comment:
@LegendreM @Kerollmops 
The implementation of `next` for the `Tokenizer` is in charge of selecting the appropriate tokenizer to use to the section of text. There are two way to go about this:
- Hardcoding the selection in a `match` => straightforward implementation
- Using a map that links scripts to tokenizer. This would allow for user configurable choice of tokenizer in the future, (wasm, ngram...) for specific scripts.